### PR TITLE
skip the processing of pflash vars file when it's not set

### PR DESCRIPTION
**Pull request by XueqiangWei**
For OVMF with SEV-ES support and OVMF with TDX support, the vm can be booted without vars file. Add a workaround, skip the processing of pflash vars file when ovmf_vars_files =.

Signed-off-by: Xueqiang Wei <xuwei@redhat.com>
ID: 1390

### CODE REVIEW

**Comment by XueqiangWei** [comment]
Tested with the code file and vars file, the results were passed.

 (1/2) Host_RHEL.m9.u4.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_local.q35: STARTED
 (1/2) Host_RHEL.m9.u4.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_local.q35: PASS (157.80 s)
 (2/2) Host_RHEL.m9.u4.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_remote.q35: STARTED
 (2/2) Host_RHEL.m9.u4.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_remote.q35: PASS (187.69 s)

 (1/6) Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x8
6_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_insta
ll.aio_threads.q35: STARTED
 (1/6) Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x8
6_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_insta
ll.aio_threads.q35: PASS (1117.93 s)
 (2/6) Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x8
6_64.io-github-autotest-qemu.boot.q35: STARTED
 (2/6) Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x8
6_64.io-github-autotest-qemu.boot.q35: PASS (72.27 s)
 (3/6) signed.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (3/6) signed.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (1498.82 s)
 (4/6) signed.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (4/6) signed.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (1767.79 s)
 (5/6) unsigned.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (5/6) unsigned.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (1441.90 s)
 (6/6) unsigned.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (6/6) unsigned.Host_RHEL.m9.u4.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (2220.95 s)


Tested with code file and without vars file, the result was passed.
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.boot.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (101.67 s)

ovmf_path = /usr/share/edk2/ovmf
ovmf_code_filename = OVMF.inteltdx.fd
ovmf_vars_filename =

qemu command line:
-blockdev '{"node-name": "file_ovmf_code", "driver": "file", "filename": "/usr/share/edk2/ovmf/OVMF.inteltdx.fd", "auto-read-only": true, "discard": "unmap"}' \
-blockdev '{"node-name": "drive_ovmf_code", "driver": "raw", "read-only": true, "file": "file_ovmf_code"}' \
-machine q35,pflash0=drive_ovmf_code,memory-backend=mem-machine_mem \

ovmf_path = /usr/share/edk2/ovmf
ovmf_code_filename = OVMF.amdsev.fd
ovmf_vars_filename =

-blockdev '{"node-name": "file_ovmf_code", "driver": "file", "filename": "/usr/share/edk2/ovmf/OVMF.amdsev.fd", "auto-read-only": true, "discard": "unmap"}' \
-blockdev '{"node-name": "drive_ovmf_code", "driver": "raw", "read-only": true, "file": "file_ovmf_code"}' \
-machine q35,pflash0=drive_ovmf_code,memory-backend=mem-machine_mem \

**Comment by XueqiangWei** [comment]
@zhencliu @YongxueHong @zixi-chen As we discussed before, add the workaround, could you please help review it? Many thanks.

**Comment by zixi-chen** [comment]
Test with sev kernelHashes install case pass. 
profile: 
[after_headers]
mem = 4096
vm_secure_guest_type = sev
vm_sev_cbitpos = 51
vm_sev_reduced_phys_bits = 1
vm_sev_policy = 7
vm_sev_kernel_hashes = on
ovmf:
    ovmf_path = /usr/share/edk2/ovmf
    ovmf_code_filename = OVMF.amdsev.fd
    ovmf_vars_filename =

 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (773.51 s)

**Comment by zhenyzha** [comment]
LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

` (1/1) Host_RHEL.m9.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.aarch64.page_4k.io-github-autotest-qemu.boot.arm64-pci: PASS (80.99 s)`

@PaulYuuu Could you pls help review when you are free. thanks.

**Review by zhenyzha — APPROVED** [review]

**Comment by PaulYuuu** [comment]
For this workaround, I am thinking about more. Actually, code file and image name have no relevance, but we have https://github.com/avocado-framework/avocado-vt/blob/5e7f791a4222236552f4f024cc67297ae1cc3424/virttest/qemu_devices/qcontainer.py#L1084-L1085

So if possible, make all of them can be added separately, this makes pflash more flexible.
```python
if firmware_path:
    ...
    ...
    if pflash_code_filename:
        ...
        ...
```

@YongxueHong @zhencliu @XueqiangWei, what do you think?

**Comment by XueqiangWei** [comment]
@PaulYuuu @YongxueHong As we discussed, this patch just handle "skip the processing of pflash vars file when it's not set". The next phase, I will refactoring and optimizing this part of the code. 
Could you please help double check this patch? Thanks.

File: virttest/qemu_devices/qcontainer.py [hunk 1]
```diff
@@ -1094,42 +1094,52 @@ def pflash_handler(firmware_name, machine_params):
...
+                        pflash_vars_src_path)["format"]
+
+                    # To ignore the influence from backends
+                    first_image = images[0]
+                    img_params = params.object_params(first_image)
+                    img_params["backing_chain"] = "no"
+                    img_obj = qemu_storage.QemuImg(img_params,
+                                                   current_data_dir,
+                                                   first_image)
+                    img_info = json.loads(img_obj.info(True, "json"))
```
[the remark this answers is not in the document]
Comment (XueqiangWei) [reply virttest/qemu_devices/qcontainer.py]: Updated, thanks.

**Comment by XueqiangWei** [comment]
Updated and retested with the code file and vars file, the results were passed.

 (1/2) Host_RHEL.m9.u3.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_local.q35: STARTED
 (1/2) Host_RHEL.m9.u3.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_local.q35: PASS (298.94 s)
 (2/2) Host_RHEL.m9.u3.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_remote.q35: STARTED
 (2/2) Host_RHEL.m9.u3.ovmf.ceph.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.remote_image_snapshot.to_remote.q35: PASS (193.98 s)
 (1/6) Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-aut
otest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/6) Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (597.23 s)
 (2/6) Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.boot.q35: STARTED
 (2/6) Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (91.76 s)
 (3/6) signed.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (3/6) signed.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (966.48 s)
 (4/6) signed.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (4/6) signed.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (1247.17 s)
 (5/6) unsigned.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (5/6) unsigned.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (1011.46 s)
 (6/6) unsigned.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: STARTED
 (6/6) unsigned.Host_RHEL.m9.u3.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.uefi_secureboot.q35: PASS (1210.30 s)


Tested with code file and without vars file, the result was passed.
 (1/1) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.boot.q35: STARTED
 (1/1) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (37.64 s)

ovmf_path = /usr/share/edk2/ovmf
ovmf_code_filename = OVMF.inteltdx.fd
ovmf_vars_filename =

qemu command line:
-blockdev '{"node-name": "file_ovmf_code", "driver": "file", "filename": "/usr/share/edk2/ovmf/OVMF.inteltdx.fd", "auto-read-only": true, "discard": "unmap"}'
-blockdev '{"node-name": "drive_ovmf_code", "driver": "raw", "read-only": true, "file": "file_ovmf_code"}'
-machine q35,pflash0=drive_ovmf_code,memory-backend=mem-machine_mem \

ovmf_path = /usr/share/edk2/ovmf
ovmf_code_filename = OVMF.amdsev.fd
ovmf_vars_filename =

-blockdev '{"node-name": "file_ovmf_code", "driver": "file", "filename": "/usr/share/edk2/ovmf/OVMF.amdsev.fd", "auto-read-only": true, "discard": "unmap"}'
-blockdev '{"node-name": "drive_ovmf_code", "driver": "raw", "read-only": true, "file": "file_ovmf_code"}'
-machine q35,pflash0=drive_ovmf_code,memory-backend=mem-machine_mem \

@YongxueHong Could you please help review it again? Many thanks.

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Comment by YongxueHong** [comment]
Hi @PaulYuuu , @zhencliu , @zixi-chen , @zhenyzha 
Would you like to have a check since there is a little updating?
Thanks.

**Review by zhencliu — APPROVED** [review]

**Comment by XueqiangWei** [comment]
According to Zhenchao's feedback, update the commit message, add "qcontainer:" in commit header. Thanks.

**Review by PaulYuuu — APPROVED** [review]

**Comment by zhenyzha** [comment]
ack

```
 (1/2) Host_RHEL.m9.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.aarch64.page_4k.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.arm64-pci: STARTED
 (1/2) Host_RHEL.m9.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.aarch64.page_4k.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.arm64-pci: PASS (543.16 s)
 (2/2) Host_RHEL.m9.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.aarch64.page_4k.io-github-autotest-qemu.boot.arm64-pci: STARTED
 (2/2) Host_RHEL.m9.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.aarch64.page_4k.io-github-autotest-qemu.boot.arm64-pci: PASS (40.67 s)
```

**Comment by zixi-chen** [comment]
@XueqiangWei should we update sev config in base.cfg and firmware config in guest-hw.cfg about the new change? 

By the way, I verified this with sev, it looks good to me.
profile:
[after_headers]
ovmf:
    ovmf_path = /usr/share/edk2/ovmf
    ovmf_code_filename = OVMF.amdsev.fd
    ovmf_vars_filename =
# python3 ConfigTest.py --testcase=sev_dhcert_boot --guestname=RHEL.9.4.0   --platform=x86_64 --clone=no --profile=profile
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.sev_dhcert_boot.q35: PASS (80.19 s)

**Comment by YongxueHong** [comment]
Thanks all, let's merge it.

**Pull request merged**